### PR TITLE
Add direct_support keys in `core_eia860__scd_generators_energy_storage` as foreign keys to `core_eia__entity_generators`

### LINKED ISSUE #3695: Add the FK relationship to the new EIA860 storage direct support plant/generators
There are new sets of plant_id's and generator id's in the `core_eia860__scd_generators_energy_storage` table which are all about whether a storage generator supports another generator. Goal of this issue is to add these new id's into special FK relationships. It is possible that we will need to harvest these new ID's in order to actually have all of the ID's in the harvested generators table. Another possible barrier is that there definitely nulls in these columns. Is it okay to have nulls in a column with a foreign key relationship to a table that doesn't have any nulls?

These are the columns at question:
```
"plant_id_eia_direct_support_1",
"generator_id_direct_support_1",
"plant_id_eia_direct_support_2",
"generator_id_direct_support_2",
"plant_id_eia_direct_support_3",
"generator_id_direct_support_3",
```
Running `pudl_check_fks` without the new alembic schema produces the following errors:
```
pudl.etl.check_foreign_keys.ForeignKeyErrors: Foreign key error for table: core_eia860__scd_generators_energy_storage -- core_eia__entity_generators (plant_id_eia, generator_id, plant_id_eia, generator_id, plant_id_eia, generator_id) -- on rows [1523 1548 1576 1589 1596 1727 1728 1795 1796 1827 1864 1865 1879 1892
 1895 1909 1920 1922 1934 1943 1946 1977 2028 2046 2059 2069 2088 2101
 2102 2105 2108 2109 2110 2111 2112 2116 2125 2126 2133 2143 2224 2236
 2315 2316 2317 2408 2448]
```

Which have the following values:

|      | plant_id_eia_direct_support_1 | generator_id_direct_support_1 | plant_id_eia_direct_support_2 | generator_id_direct_support_2 | plant_id_eia_direct_support_3 | generator_id_direct_support_3 |
|-----:|------------------------------:|------------------------------:|------------------------------:|------------------------------:|------------------------------:|------------------------------:|
| 1523 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1548 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1576 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1589 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1596 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1727 |                       62975.0 |                         SYNLB |                           NaN |                          None |                           NaN |                          None |
| 1728 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1795 |                       63490.0 |                           WIL |                           NaN |                          None |                           NaN |                          None |
| 1796 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1827 |                       64185.0 |                         DRAPV |                           NaN |                          None |                           NaN |                          None |
| 1864 |                       64436.0 |                           WLB |                           NaN |                          None |                           NaN |                          None |
| 1865 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1879 |                       64520.0 |                          SOL1 |                           NaN |                          None |                           NaN |                          None |
| 1892 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1895 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1909 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1920 |                       64852.0 |                         DAGGB |                       64852.0 |                         DAGGP |                           NaN |                          None |
| 1922 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1934 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 1943 |                       65055.0 |                         77FPV |                           NaN |                          None |                           NaN |                          None |
| 1946 |                       65058.0 |                         WAIBA |                       65058.0 |                         WAIPV |                           NaN |                          None |
| 1977 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2028 |                       66076.0 |                         20409 |                           NaN |                          None |                           NaN |                          None |
| 2046 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2059 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2069 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2088 |                       67457.0 |                         SHPV1 |                           NaN |                          None |                           NaN |                          None |
| 2101 |                       61169.0 |                         60798 |                           NaN |                          None |                           NaN |                          None |
| 2102 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2105 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2108 |                       61720.0 |                            30 |                           NaN |                          None |                           NaN |                          None |
| 2109 |                       61722.0 |                            32 |                           NaN |                          None |                           NaN |                          None |
| 2110 |                       61752.0 |                            49 |                           NaN |                          None |                           NaN |                          None |
| 2111 |                       61807.0 |                            66 |                           NaN |                          None |                           NaN |                          None |
| 2112 |                       62494.0 |                         SCSOL |                           NaN |                          None |                           NaN |                          None |
| 2116 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2125 |                       63541.0 |                         63257 |                           NaN |                          None |                           NaN |                          None |
| 2126 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2133 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2143 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2224 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2236 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2315 |                       65550.0 |                         64921 |                           NaN |                          None |                           NaN |                          None |
| 2316 |                       65550.0 |                         65789 |                           NaN |                          None |                           NaN |                          None |
| 2317 |                       65550.0 |                          None |                           NaN |                          None |                           NaN |                          None |
| 2408 |                           NaN |                          None |                           NaN |                          None |                           NaN |                          None |
| 2448 |                       67282.0 |                          5654 |                           NaN |                          None |                           NaN |                          None |

```[tasklist]
### Tasks
- [x] Add new `foreign_key_rules` into `core_eia__entity_generators` for each of  these direct support column names for the plant/gen ids. (examples of other tables that do this are: `core_eia__entity_utilities`, `core_eia860__scd_utilities`)
- [x] Regenerate the alembic schema
- [x] test the relationships w/ `pudl_check_fks`
- [x] fix plant ID dtypes
- [x] Rule out incorrectly written string nulls in SQLite DB by directly querying in DB Browser
- [x] debug NaNs / Nones failing the FK check
```

### PR DESCRIPTION
**Pull request by e-belfer**
# Overview

Closes #3695.

What problem does this address?
Ensures FK constraints addressed between the new storage table and the harvested EIA generator and plants IDs.

What did you change?
- Make `plant_id_eia_direct_support_x` fields into integers
- Add foreign key constraint
- Harvest plant/gen ID combinations
- Drop 5 new and totally empty plant/generator/report year records from `_out_eia__yearly_generators` to avoid introducing junk records into our output tables.

# Testing

How did you make sure this worked? How can a reviewer verify this?
Update alembic, regenerate changed assets and run `pudl_check_fks`. Manually inspect the generator entities table to make sure that the IDs from the storage table are present.

# Question for reviewer
I'm almost certain that rows 2315-6 in `core_eia860__scd_generators_energy_storage` are reporting a second plant ID in the generator support 1 column. Based on the non-existence of those generator IDs in all other EIA data, and the fact that they're valid plant IDs. Should we do anything about this, or accept the introduction of bad generator IDs into our harvesting process?

_Decision_: We're dropping these records in our `out` tables when there is no other data associated with them, so if they ever become real records we'll get an assertion error.

```[tasklist]
# To-do list
- [x] Review the PR yourself and call out any questions or issues you have
- [x] Clean up alembic migrations
- [x] Check for unmapped plant IDs
- [x] Update min/max row counts
- [x] Drop the "known to be empty" ids in the output table, and add an assertion if they are not. Only drop the selected IDs.
- [x] Kick off a build and debug any test failures
- [x] Update min-max row counts again after 923 PR merges
- [x] Resolve null row in `mcoe_all_gens`
```

### CODE REVIEW

**Review by cmgosnell — CHANGES_REQUESTED** [review]
amazing!!

one probably nit about boiler entity harvesting.

and one completely non-blocking request to combine the three migrations into one. I think you could do this by `alembic downgrade head-3` -> delete these three new migrations -> `alembic revision --autogenerate -m "Add my cool new direct support FKs"` -> `alembic upgrade head`

File: src/pudl/metadata/resources/__init__.py [hunk 1]
```diff
@@ -265,6 +280,13 @@
             "standard_so2_percent_scrubbed",
             "data_maturity",
         ],
+        "mapped_schemas": [
+            {
+                "plant_id_eia_direct_support_1": "plant_id_eia",
+                "plant_id_eia_direct_support_2": "plant_id_eia",
+                "plant_id_eia_direct_support_3": "plant_id_eia",
+            }
+        ],
```
[the remark this answers is not in the document]
Comment (e-belfer) [reply src/pudl/metadata/resources/__init__.py:289]: When the `boiler_id` is null any newly added `plant_id_eia` records this won't show up in the table, so no, you're correct. I will remove this.

**Review by cmgosnell — DISMISSED** [review]
ty for the bb updates!

two last requests: did you check to see if this new harvesting of plant id's grabbed completely new plants? because if so we will need to add them into the pudl id mapping process. [docs](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/existing_data_updates.html#connect-datasets)

it may also change the min/max rows of the generator-based tables!

**Comment by e-belfer** [comment]
@cmgosnell I expect there might be some additional row count failures that are caused by the harvesting, but I'm unable to run them locally. I'd like to just merge them into the nightly builds and patch a fix on Monday to update any remaining row count failures.

**Comment by zaneselvans** [comment]
@e-belfer I think you're done for the day/week and I need to re-run my ETL locally anyway so I can give this a try and update any row counts that are off so we have a chance at getting a good build by Monday.

**Comment by zaneselvans** [comment]
### Excessively null generators

Running the full ETL locally, there were some excessively null rows in the yearly and monthly generator output tables. I'd guess because these generator IDs only showed up in the new table, and were missing from the tables that provide the majority of the generator attributes. See error messages below.

<details><summary>Monthly/Yearly Generator Output Errors</summary>

```
dagster._core.errors.DagsterExecutionStepExecutionError: Error occurred while executing op "out_eia__yearly_generators":

The above exception was caused by the following exception:
ValueError: Found 4 Null rows in mcoe_all_gens.

  File "/Users/zane/code/catalyst/pudl/src/pudl/analysis/mcoe.py", line 209, in mcoe_generators_asset
    return mcoe_generators(
           ^^^^^^^^^^^^^^^^
  File "/Users/zane/code/catalyst/pudl/src/pudl/analysis/mcoe.py", line 704, in mcoe_generators
    ).pipe(pudl.validate.no_null_rows, df_name="mcoe_all_gens", thresh=0.9)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/pandas/core/generic.py", line 6231, in pipe
    return common.pipe(self, func, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/pandas/core/common.py", line 502, in pipe
    return func(obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zane/code/catalyst/pudl/src/pudl/validate.py", line 124, in no_null_rows
    raise ValueError(f"Found {null_rows.sum(axis='rows')} Null rows in {df_name}.")
```

```
dagster._core.errors.DagsterExecutionStepExecutionError: Error occurred while executing op "out_eia__monthly_generators":

The above exception was caused by the following exception:
ValueError: Found 4 Null rows in mcoe_all_gens.

  File "/Users/zane/code/catalyst/pudl/src/pudl/analysis/mcoe.py", line 209, in mcoe_generators_asset
    return mcoe_generators(
           ^^^^^^^^^^^^^^^^
  File "/Users/zane/code/catalyst/pudl/src/pudl/analysis/mcoe.py", line 704, in mcoe_generators
    ).pipe(pudl.validate.no_null_rows, df_name="mcoe_all_gens", thresh=0.9)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/pandas/core/generic.py", line 6231, in pipe
    return common.pipe(self, func, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zane/miniforge3/envs/pudl-dev/lib/python3.12/site-packages/pandas/core/common.py", line 502, in pipe
    return func(obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zane/code/catalyst/pudl/src/pudl/validate.py", line 124, in no_null_rows
    raise ValueError(f"Found {null_rows.sum(axis='rows')} Null rows in {df_name}.")
```

</details>

### Expected Row Counts

Not all tables were populated due to the above errors, but for those that were populated, there were no additional changes to the row counts beyond those already documented. The only tables left unpopulated are related to the EIA Plant Parts and the FERC-EIA record linkage.

### Unmapped Plant/Utility IDs

Generating this list now...

**Comment by e-belfer** [comment]
I've mapped the IDs, but these four plants are still missing a PUDL plant ID for the subsequent output tables. Need to look into this further.

**Review by aesharpe — APPROVED** [review]
Looks good! all assets materialized and there are no foreign key errors.

**Comment by e-belfer** [comment]
The integration test failure has to do with one generator (EIA plant ID 60321, generator ID PV1) which is reported in the 860 storage direct support columns in 2023 ER data but nowhere else in the fast ETL data, causing a completely empty row and the corresponding failure.

Problem is, that plant ID and another generator (EXIS2) do exist in our data in other years (retired 2019), and so the plant attributes (location, lat/lon coords, etc.) are carrying through in the full ETL run and not triggering the null row error. The generator itself never has data. So, consistent with the other generators dropped, I've added it to the `empty_generator_ids` list.

**Review by cmgosnell — APPROVED** [review]
annoying but reasonable. if totally null generators continues to be a problem it might be good to apply a fix that is less specific to particular generators. but this feels like by far the most secure way to do this so 👍🏻

**Pull request merged**